### DISCUSSION TITLE
Planning: Limit ability of synthetic nodes to take up connection slots

### DISCUSSION THREAD

**Discussion by arya2** [ZcashFoundation/zebra#7987] [Planning]
## Motivation

We want to discuss the impact of each potential fix on this vulnerability, and decide on 1-3 fixes that have the greatest impact.

### Potential Fixes

This is the list of potential fixes from issue https://github.com/ZcashFoundation/zebra/issues/7822:

- [ ] Increase the minimum number of connected peers to the initial target peerset size, so that Zebra always has a variety of peers, and the first/newest/fastest peers don't take up all the slots (also helps with #7824)
 https://github.com/ZcashFoundation/zebra/blob/7e7f989545bac06ee674547425ac0ac907cc0bce/zebra-network/src/peer_set/initialize.rs#L913
- [ ] Retry failed peers only if we've successfully connected to them, don't believe `untrusted_last_seen` https://github.com/ZcashFoundation/zebra/blob/7e7f989545bac06ee674547425ac0ac907cc0bce/zebra-network/src/meta_addr.rs#L639-L640
- [ ] Retry other failed peers (untrusted or no last seen time) if their last attempt is longer than `MAX_PEER_ACTIVE_FOR_GOSSIP` (or maybe just 10+ minutes?) but less than `MAX_RECENT_PEER_AGE` 
- [ ] #7852
- [ ] #1875
- [ ] If we're not getting any new blocks in the progress task, send a `ChangePeerConnections` request to the network to disconnect some peers
- [ ] Increase `GET_ADDR_FANOUT` to 2 or 3, to increase the diversity of the address book
- [ ] Reduce the `TimerCrawl` interval to increase the diversity of the address book

**Comment by arya2** [comment]
Is there a good way for Zebra to recognize and drop synthetic peers once they're connected?

**Reply by teor2345** [reply]
I don't think there is. Any algorithm that tries to do general detection of fake or bad peers can be defeated by this attack:
1. Start a fake peer.
2. Connect to the target peer.
3. Connect to another real peer.
4. When the target peer sends messages, proxy them to the real peer, and return its responses. Similarly, forward unsolicited messages from the real peer to the target peer.
5. Modify or introduce traffic as needed to attack the target peer.

Instead, if Zebra has a specific requirement of its peers, I suggest we rotate peers until we get peers that satisfy that requirement. For example, this suggested fix:
> If we're not getting any new blocks in the progress task, send a ChangePeerConnections request to the network to disconnect some peers

**Reply by arya2** [reply]
> When the target peer sends messages, proxy them to the real peer, and return its responses. Similarly, forward unsolicited messages from the real peer to the target peer.

Is it possible for the target to guess the likelihood that there's a proxy among a subset of peers when it's connected to the real peer by multiple proxies?

**Reply by teor2345** [reply]
> > When the target peer sends messages, proxy them to the real peer, and return its responses. Similarly, forward unsolicited messages from the real peer to the target peer.
> 
> Is it possible for the target to guess the likelihood that there's a proxy among a subset of peers when it's connected to the real peer by multiple proxies?

It might help to consider this attack as a [thought experiment](https://en.wikipedia.org/wiki/Thought_experiment), rather than an actual practical attack. The underlying assumption is that it is possible to easily emulate and modify a real peer to get around any general rule about what peers should do.

For example, you can also emulate a peer by forking software, recording and replaying responses, or creating a minimal peer implementation that satisfies the general "good peer" rules.

And as a side note, creating those kinds of general rules about peer behaviour also makes it harder to write compatible protocol implementations. And it can open the network up to other attacks that induce "bad" behaviour in real peers.

**Comment by arya2** [comment]
> Increase the minimum number of connected peers to the initial target peerset size, so that Zebra always has a variety of peers, and the first/newest/fastest peers don't take up all the slots

This slows down how quickly an adversary could fill up the connection slots because the connectors are rate limited.

This is likely a required fix depending on how much time Zebra needs to confidently recognize and drop synthetic peers.

>  Retry failed peers only if we've successfully connected to them, don't believe untrusted_last_seen

This is a required fix to prevent an easy exploit, there are alternatives, but I think we should do this one.

>  Retry other failed peers (untrusted or no last seen time) if their last attempt is longer than MAX_PEER_ACTIVE_FOR_GOSSIP (or maybe just 10+ minutes?) but less than MAX_RECENT_PEER_AGE

This may help in some cases where Zebra's nearly starved of peers, I'm not entirely sure, it depends on the other fixes, but it seems only tangentially relevant to me.

> https://github.com/ZcashFoundation/zebra/issues/7852

This is likely a required fix depending on how much time Zebra needs to confidently recognize and drop synthetic peers.

> https://github.com/ZcashFoundation/zebra/issues/1875

Benefit: This practically prevents any attempt to predict which addresses are used by Zebra's outbound connector.

Drawbacks (from the issue):
> this change risks connecting to slightly worse peers, or skipping some peers that would otherwise have been chosen for reconnection.

The drawback seems to outweigh the benefit, it's already very difficult to predict which addresses make it to the address book to be used by Zebra's outbound connector.

I think always caching some portion (in this case half) of the addresses from a `GetAddr` response, even if it's below half of the outbound connection limit would accomplish the same goal with a less-severe drawback in normal conditions.

> If we're not getting any new blocks in the progress task, send a ChangePeerConnections request to the network to disconnect some peers

This is a required fix so Zebra can recover when synthetic nodes take up all of its connections slots.

If there's a way to test/score how likely a peer is to be a synthetic node, this should happen even when Zebra *is* getting new blocks from some of its peers.

>  Increase GET_ADDR_FANOUT to 2 or 3, to increase the diversity of the address book

I think this helps by making it more likely we'll dial an address from a random peer rather than from the address book which favours addresses from some peers.

I don't see it fixing the core issue.

>  Reduce the TimerCrawl interval to increase the diversity of the address book

This will likely make it _easier_ to fill every connection slot quickly if it continues to dial on every iteration where `crawl` has sent new addresses to the address book.

We should make sure to skip dialing if there was a dial within 2-3 prior crawl iterations if we do this instead or increasing `GET_ADDR_FANOUT`.

If we do that, it should have about the same impact as the increasing the `GET_ADDR_FANOUT`.

**Reply by teor2345** [reply]
@arya2 I think it might be hard to have a discussion in response to one large comment. It might end up the same way as the discussion on the last ticket.

Here's one possible way to do it using comment threads:
- have one top-level comment that describes each suggestion
- multiple people can reply to that comment with the benefits/drawbacks/further analysis/priority
- edit the top-level comment (or the ticket) to summarise the discussion 

That way, all the relevant information for each suggestion will be grouped together.

**Comment by arya2** [comment]
Potential Fix: Increase the minimum number of connected peers to the initial target peerset size, so that Zebra always has a variety of peers, and the first/newest/fastest peers don't take up all the slots

**Reply by arya2** [reply]
This slows down how quickly an adversary could fill up the connection slots because the connectors are rate limited.

This is likely a required fix depending on how much time Zebra needs to confidently recognize and drop synthetic peers.

**Reply by teor2345** [reply]
Disadvantage: increases the load on the network from peers that aren't doing any work. This is a minor disadvantage, being vulnerable to attacks is worse.

> This is likely a required fix depending on how much time Zebra needs to confidently recognize and drop synthetic peers.

Unless there is a required fix that we are sure will identify all synthetic peers, then this fix is required.

**Comment by arya2** [comment]
Potential Fix: Retry failed peers only if we've successfully connected to them, don't believe `untrusted_last_seen`

**Reply by arya2** [reply]
This is a required fix to prevent an easy exploit, there are alternatives, but I think we should do this one.

**Reply by teor2345** [reply]
Drawbacks: peers won't be retried if Zebra starts while disconnected from the network, or becomes disconnected while trying a lot of new peers. This is a major drawback that will significantly impact Zebra's connectivity.

Alternative: Instead, I suggest we retry peers without successful connections last.
Analysis Needed: Is this the existing implementation, or do we need to make changes?

**Reply by teor2345** [reply]
The existing behaviour is to order peers without successful connections last, with the standard 2 minute reconnection delay:
https://github.com/ZcashFoundation/zebra/blob/193fd427fb1717a89ceb450796f93275fb215e29/zebra-network/src/meta_addr.rs#L1122-L1132

So do we need to change anything here?

**Comment by arya2** [comment]
Potential Fix: Retry other failed peers (untrusted or no last seen time) if their last attempt is longer than `MAX_PEER_ACTIVE_FOR_GOSSIP` (or maybe just 10+ minutes?) but less than `MAX_RECENT_PEER_AGE`

**Reply by arya2** [reply]
This may help in some cases where Zebra's nearly starved of peers, I'm not entirely sure, it depends on the other fixes, but it seems only tangentially relevant to me.

**Reply by teor2345** [reply]
Is this an alternative or a mitigation for the drawbacks of "Retry failed peers only if we've successfully connected to them"?

Benefits: Possible mitigation for other drawbacks.
Drawbacks: Delays could impact connectivity, we'd need to track another timer, hard to analyse.

Analysis Needed: Is the existing behaviour to have a delay, never retry, or retry last with no delay?

**Reply by teor2345** [reply]
The existing behaviour is to retry last with the standard 2 minute delay, see:
https://github.com/ZcashFoundation/zebra/discussions/7987#discussioncomment-7676065

**Reply by teor2345** [reply]
I don't think we need to make any changes here.

**Comment by arya2** [comment]
Potential Fix: 
- https://github.com/ZcashFoundation/zebra/issues/7852

**Reply by arya2** [reply]
This is likely a required fix depending on how much time Zebra needs to confidently recognize and drop synthetic peers.

**Reply by teor2345** [reply]
We already delay reconnections to the same IP for around 2 minutes, unless that IP is pushed out of the address book by new gossiped peers.

Benefits: Possible mitigation for rapid peer reconnections in a large address book.
Drawbacks: Duplicates existing behaviour but in a different way, we'd need to track another timer, hard to analyse.

Alternatives:
- Increase the size of the address book
- Increase the existing per-IP reconnection delay

Analysis Needed:
- Is the address book limit ever reached after our recent changes? How often, and how many peers are removed? Are they valid peers where a reconnection could cause performance issues or network load?
- Is the per-IP reconnection delay long enough? (2 minutes means there will be one connection per second if 120 peers behave this way.)

**Reply by teor2345** [reply]
I think the root cause/fix here is having diverse peers to connect to. Delaying the connection doesn't change which peer we will connect to next, so I'm not sure if this fix will help.

**Reply by teor2345** [reply]
There is an attack where an attacker tries to add lots of peers, fills the address book, and pushes their own IPs out of the address book, then adds their IPs again, and then has us reconnect within less than our 2 minute limit.

This attack isn't viable because:
- Zebra controls when peers are added to the address book, and limits the amount and frequency of peers being added
- Zebra limits outbound connection rates, it takes 8 minutes for Zebra to connect to a full address book, but the individual IP limit is 2 minutes

So it is actually better for the peer to stay in the address book.

**Reply by teor2345** [reply]
I don't think this change is guaranteed to help, and it might hurt security in other ways.

**Comment by arya2** [comment]
Potential Fix: 
- https://github.com/ZcashFoundation/zebra/issues/1875

**Reply by arya2** [reply]
Benefit: This practically prevents any attempt to predict which addresses are used by Zebra's outbound connector.

Drawbacks (from the issue):
> this change risks connecting to slightly worse peers, or skipping some peers that would otherwise have been chosen for reconnection.

The drawback seems to outweigh the benefit, it's already very difficult to predict which addresses make it to the address book to be used by Zebra's outbound connector.

I think always caching some portion (in this case half) of the addresses from a `GetAddr` response, even if it's below half of the outbound connection limit would accomplish the same goal with a less-severe drawback in normal conditions.

**Reply by teor2345** [reply]
Since #7824, peer are chosen at random from gossiped addresses.

> I think always caching some portion (in this case half) of the addresses from a GetAddr response, even if it's below half of the outbound connection limit would accomplish the same goal with a less-severe drawback in normal conditions.

We could make that randomness apply to more peer gossips, but whatever limit we choose, there will be some size of gossip that remote peers will be able to exploit. (In this case, 37 or less.)

But that wouldn't matter if we randomised reconnections themselves.

Once peers are in the address book, it is easy to predict or exploit their connection order, because we use a fixed ordering for reconnections, based on fields controlled by remote peers:
https://github.com/ZcashFoundation/zebra/blob/193fd427fb1717a89ceb450796f93275fb215e29/zebra-network/src/address_book.rs#L590-L602
And MetaAddr:
https://github.com/ZcashFoundation/zebra/blob/193fd427fb1717a89ceb450796f93275fb215e29/zebra-network/src/meta_addr.rs#L1089-L1171
And PeerAddrState:
https://github.com/ZcashFoundation/zebra/blob/193fd427fb1717a89ceb450796f93275fb215e29/zebra-network/src/meta_addr.rs#L132-L145

**Reply by teor2345** [reply]
Candidates for replacement with a random order
```
     .then(newer_untrusted_last_seen) 
     .then(larger_services) 
     .then(ip_tie_breaker) 
     .then(port_tie_breaker) 
```

**Reply by teor2345** [reply]
Replace the tie-breakers with a `random_reconnection_order` field that is randomised once when the peer is created. This field is only effectively used for the first connection attempt, after that the order is based on the response/failure times.

We should document that if a peer gets a higher random value, then it might never be attempted, because new peers are added with different values.

`random_reconnection_order` should be a `u16` so that almost all peers have a unique value. We won't need any tie-breakers, untrusted fields, and we don't depend on `HashMap` order.

**Reply by teor2345** [reply]
This seems like a useful fix, it might also be required, I'm not sure.

**Comment by arya2** [comment]
Potential Fix: If we're not getting any new blocks in the progress task, send a ChangePeerConnections request to the network to disconnect some peers

**Reply by arya2** [reply]
This is a required fix so Zebra can recover when synthetic nodes take up all of its connections slots.

If there's a way to test/score how likely a peer is to be a synthetic node, this should happen even when Zebra *is* getting new blocks from some of its peers.

**Reply by teor2345** [reply]
What are the benefits and drawbacks of this fix?

**Reply by teor2345** [reply]
Benefit: we will probably need something like this anyway for:
- #4545 

(But with a setting, or a change to the peer set / handshaker to drop new peers which are on outdated versions.)

**Reply by teor2345** [reply]
Do we do this when we're not getting anything, or also when we're failing validation?
Do we want a separate ticket for the mempool using this mechanism?

**Reply by teor2345** [reply]
> Do we do this when we're not getting anything, or also when we're failing validation?

Doing this when we haven't had any new blocks for more than `validation timeout + restart timeout + time for the first block to verify` will address both cases.

> Do we want a separate ticket for the mempool using this mechanism?

Yes, it isn't part of this fix, and neither is #4545.

**Reply by teor2345** [reply]
This seems like it is required.

**Comment by arya2** [comment]
Potential Fix: Increase `GET_ADDR_FANOUT` to 2 or 3, to increase the diversity of the address book

**Reply by arya2** [reply]
I think this helps by making it more likely we'll dial an address from a random peer rather than from the address book which favours addresses from some peers.

I don't see it fixing the core issue.

**Reply by teor2345** [reply]
I think it is required to fix this issue, so that a single peer is less likely to be able to supply all the best and latest addresses for connection attempts:
https://github.com/ZcashFoundation/zebra/discussions/7987#discussioncomment-7676054

**Comment by arya2** [comment]
Potential Fix:  Reduce the `TimerCrawl` interval to increase the diversity of the address book

**Reply by arya2** [reply]
This will likely make it _easier_ to fill every connection slot quickly if it continues to dial on every iteration where `crawl` has sent new addresses to the address book.

We should make sure to skip dialing if there was a dial within 2-3 prior crawl iterations if we do this instead or increasing `GET_ADDR_FANOUT`.

If we do that, it should have about the same impact as the increasing the `GET_ADDR_FANOUT`.

**Reply by teor2345** [reply]
> This will likely make it _easier_ to fill every connection slot quickly if it continues to dial on every iteration where `crawl` has sent new addresses to the address book.

This is a drawback, but a benefit is that a single peer (or small group of peers) are less likely to supply all the best and latest addresses for connection attempts over the span of a minute or two, particularly when Zebra is idle.

That's related to this issue:
https://github.com/ZcashFoundation/zebra/discussions/7987#discussioncomment-7676054

**Reply by teor2345** [reply]
> We should make sure to skip dialing if there was a dial within 2-3 prior crawl iterations if we do this instead or increasing `GET_ADDR_FANOUT`.
> 
> If we do that, it should have about the same impact as the increasing the `GET_ADDR_FANOUT`.

This is an alternative fix.

What are the benefits and drawbacks?

Here is one drawback: if we skip dials, a single peer (or small group of peers) is much more likely to supply all the best and latest addresses for successful connection over the span of a minute or two, particularly when Zebra is idle. This seems like a major impact.

**Reply by teor2345** [reply]
I think that it is better to increase the fanout, because then each address book update has a more diverse set of peers.

I don't think this fix is guaranteed to be useful, and it seems to have some drawbacks.

**Comment by teor2345** [comment]
@arya2 @mpguerra I suggest we make issues for the things we're sure of, and leave the discussion of the rest until the team has more capacity.
